File: templates/problems/09_heap/02_last_stone_weight/typescript/last_stone_weight.test.ts
import { describe, it, expect } from "vitest";
import { lastStoneWeight } from "./last_stone_weight";

describe("lastStoneWeight", () => {
  it("example from problem", () => {
    expect(lastStoneWeight([2, 7, 4, 1, 8, 1])).toBe(1);
  });

  it("single stone", () => {
    expect(lastStoneWeight([1])).toBe(1);
  });

  it("two equal stones", () => {
    expect(lastStoneWeight([3, 3])).toBe(0);
  });

  it("two different stones", () => {
    expect(lastStoneWeight([1, 5])).toBe(4);
  });

  it("all same weight", () => {
    expect(lastStoneWeight([2, 2, 2, 2])).toBe(0);
  });

  it("descending weights", () => {
    expect(lastStoneWeight([10, 5, 3, 1])).toBe(1);
  });

  it("three stones", () => {
    expect(lastStoneWeight([3, 7, 2])).toBe(2);
  });
});

File: templates/problems/09_heap/02_last_stone_weight/go/last_stone_weight_test.go
package last_stone_weight

import "testing"

func TestLastStoneWeight(t *testing.T) {
	tests := []struct {
		name   string
		stones []int
		want   int
	}{
		{
			name:   "example from problem",
			stones: []int{2, 7, 4, 1, 8, 1},
			want:   1,
		},
		{
			name:   "single stone",
			stones: []int{1},
			want:   1,
		},
		{
			name:   "two equal stones",
			stones: []int{3, 3},
			want:   0,
		},
		{
			name:   "two different stones",
			stones: []int{1, 5},
			want:   4,
		},
		{
			name:   "all same weight",
			stones: []int{2, 2, 2, 2},
			want:   0,
		},
		{
			name:   "descending weights",
			stones: []int{10, 5, 3, 1},
			want:   1,
		},
		{
			name:   "three stones",
			stones: []int{3, 7, 2},
			want:   2,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := lastStoneWeight(tt.stones)
			if got != tt.want {
				t.Errorf("lastStoneWeight(%v) = %d, want %d", tt.stones, got, tt.want)
			}
		})
	}
}
